src/ucp/core/ucp_worker.c

```diff
@@ -2468,33 +2469,39 @@ static void ucp_worker_discard_uct_ep_complete(ucp_request_t *req)
     ucp_ep_refcount_remove(ucp_ep, discard);
 }
 
-static unsigned ucp_worker_discard_uct_ep_destroy_progress(void *arg)
+static unsigned ucp_worker_discard_uct_ep_destroy_once(ucp_request_t *req)
 {
-    ucp_request_t *req        = (ucp_request_t*)arg;
     uct_ep_h uct_ep           = req->send.discard_uct_ep.uct_ep;
     ucp_rsc_index_t rsc_index = req->send.discard_uct_ep.rsc_index;
     ucp_ep_h ucp_ep           = req->send.ep;
-    ucp_worker_h worker       = ucp_ep->worker;
-    khiter_t iter;
 
     ucp_trace_req(req, "destroy uct_ep=%p", uct_ep);
 
     req->send.discard_uct_ep.cb_id = UCS_CALLBACKQ_ID_NULL;
-
-    UCS_ASYNC_BLOCK(&worker->async);
-    iter = kh_get(ucp_worker_discard_uct_ep_hash, &worker->discard_uct_ep_hash,
-                  uct_ep);
-    if (iter == kh_end(&worker->discard_uct_ep_hash)) {
-        ucs_fatal("no %p UCT EP in the %p worker hash of discarded UCT EPs",
-                  uct_ep, worker);
-    }
-
     ucp_ep_unprogress_uct_ep(ucp_ep, uct_ep, rsc_index);
     uct_ep_destroy(uct_ep);
+
+    req->send.discard_uct_ep.state = UCP_WORKER_DISCARD_UCT_EP_FINISHED;
     ucp_worker_discard_uct_ep_complete(req);
 
-    ucs_assert(kh_value(&worker->discard_uct_ep_hash, iter) == req);
-    kh_del(ucp_worker_discard_uct_ep_hash, &worker->discard_uct_ep_hash, iter);
+    return 1;
+}
+
+static unsigned ucp_worker_discard_uct_ep_destroy_progress(void *arg)
+{
+    ucp_request_t *req  = (ucp_request_t*)arg;
+    ucp_worker_h worker = req->send.ep->worker;
+
+    UCS_ASYNC_BLOCK(&worker->async);
+    kh_foreach_value(&worker->discard_uct_ep_hash, req, {
+        if (req->send.discard_uct_ep.state ==
+            UCP_WORKER_DISCARD_UCT_EP_FLUSHED) {
+            kh_del(ucp_worker_discard_uct_ep_hash,
+                   &worker->discard_uct_ep_hash, __i);
+            ucp_worker_discard_uct_ep_destroy_once(req);
+        }
+    });
+    worker->discard_uct_ep_comp = 0;
```

Comment:
please do `worker->discard_uct_ep_comp--` in the loop then assert that it's `0` in the end

Reply:
done